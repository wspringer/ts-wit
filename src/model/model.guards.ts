import {
  TypeDef,
  AliasDef,
  EnumDef,
  VariantDef,
  RecordDef,
  Ty,
  TypeRef,
  SimpleType,
  OptionType,
  ListType,
  ResultType,
  TupleType,
} from "./model.types";

export function isAliasDef<M>(typeDef: TypeDef<M>): typeDef is AliasDef<M> {
  return typeDef.kind === "alias";
}

export function isEnumDef<M>(typeDef: TypeDef<M>): typeDef is EnumDef<M> {
  return typeDef.kind === "enum";
}

export function isVariantDef<M>(typeDef: TypeDef<M>): typeDef is VariantDef<M> {
  return typeDef.kind === "variant";
}

export function isRecordDef<M>(typeDef: TypeDef<M>): typeDef is RecordDef<M> {
  return typeDef.kind === "record";
}

export function isTypeRef<M>(ty: Ty<M>): ty is TypeRef<M> {
  return typeof ty === "object" && "name" in ty && !("kind" in ty);
}

export function isSimpleType<M>(ty: Ty<M>): ty is SimpleType {
  return (
    typeof ty === "string" &&
    [
      "s8",
      "s16",
      "s32",
      "s64",
      "u8",
      "u16",
      "u32",
      "u64",
      "f32",
      "f64",
      "bool",
      "string",
      "char",
    ].includes(ty)
  );
}

export function isOptionType<M>(ty: Ty<M>): ty is OptionType<M> {
  return typeof ty === "object" && "kind" in ty && ty.kind === "option";
}

export function isListType<M>(ty: Ty<M>): ty is ListType<M> {
  return typeof ty === "object" && "kind" in ty && ty.kind === "list";
}

export function isResultType<M>(ty: Ty<M>): ty is ResultType<M> {
  return typeof ty === "object" && "kind" in ty && ty.kind === "result";
}

export function isTupleType<M>(ty: Ty<M>): ty is TupleType<M> {
  return typeof ty === "object" && "items" in ty && ty.kind === "tuple";
}
