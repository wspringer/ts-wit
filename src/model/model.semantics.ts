import { WitSemantics } from "../grammar.ohm-bundle";
import {
  AliasDef,
  EnumDef,
  FlagsDef,
  Func,
  InterfaceDef,
  ListType,
  NestedPackage,
  OptionType,
  Package,
  Param,
  RecordDef,
  RecordField,
  ResultType,
  SimpleType,
  Ty,
  TypeDef,
  VariantCase,
  VariantDef,
  Wit,
} from "./model.types";
import { grammar } from "../grammar";

type Item<K, T> = {
  kind: K;
  boxed: T;
};

function defineModel(semantics: WitSemantics) {
  semantics.addOperation<any>("toModel", {
    File(packageDecl, semicolon, packageItems): Wit {
      const items: Item<any, any>[] = packageItems.children.map((child) =>
        child.toModel()
      );
      const interfaces: InterfaceDef[] = items
        .filter((item) => item.kind === "interface")
        .map((item) => item.boxed);
      const packages: NestedPackage[] = items
        .filter((item) => item.kind === "package")
        .map((item) => item.boxed);

      return {
        name: packageDecl.children[0]?.toModel(),
        uses: [],
        packages,
        interfaces,
        worlds: [],
      };
    },

    PackageDecl(package_, name): string {
      return name.sourceString;
    },

    InterfaceItem(
      interface_,
      name,
      openBrace,
      interfaceItems,
      closeBrace
    ): Item<"interface", InterfaceDef> {
      const items = interfaceItems.children.map((child) => child.toModel());
      const functions: Func[] = items
        .filter((item) => item.kind === "func")
        .map((item) => item.boxed);
      const typeDefs: TypeDef[] = items
        .filter((item) => item.kind === "typeDef")
        .map((item) => item.boxed);
      return {
        kind: "interface",
        boxed: {
          name: name.sourceString,
          functions,
          typeDefs,
        },
      };
    },

    InterfaceItems(gate, definition): InterfaceDef {
      return definition.toModel();
    },

    FuncItem(ident, colon, funcType, semicolon): Item<"func", Func> {
      const { params, result } = funcType.toModel();
      const boxed: Func = {
        name: ident.sourceString,
        params,
        result,
      };
      return {
        kind: "func",
        boxed,
      };
    },

    FuncType(async_, func, params, result): Omit<Func, "name"> {
      return {
        params: params.toModel(),
        // params: [],
        result: result?.children[0]?.toModel(),
      };
    },

    ParamList(openParen, params, closeParen): Param[] {
      return params.toModel();
    },

    NamedTypeList(params): Param[] {
      params.children[0].children.length; //?
      return params.children[0].children.map((child) => child.toModel());
    },

    NamedType(name, colon, type): Param {
      return {
        name: name.toModel(),
        type: type.toModel(),
      };
    },

    ResultList(arrow, result): Ty {
      return result.toModel();
    },

    ListType(list, openAngle, type, closeAngle): ListType {
      return {
        kind: "list",
        type: type.toModel(),
      };
    },

    OptionType(option, openAngle, type, closeAngle): OptionType {
      return {
        kind: "option",
        type: type.toModel(),
      };
    },

    ResultType(result, openAngle, ok, comma, error, closeAngle): ResultType {
      return {
        kind: "result",
        ok: ok.toModel(),
        error: error.children[1]?.toModel(),
      };
    },

    SimpleType(type): SimpleType {
      return type.sourceString as SimpleType;
    },

    EnumItem(
      enum_,
      ident,
      openBrace,
      enumItems,
      comma,
      closeBrace
    ): Item<"typeDef", EnumDef> {
      return {
        kind: "typeDef",
        boxed: {
          kind: "enum",
          name: ident.sourceString,
          cases: enumItems
            .asIteration()
            .children.map((child) => child.toModel()),
        },
      };
    },

    EnumCase(ident): string {
      return ident.sourceString;
    },

    VariantItem(
      variant,
      ident,
      openBrace,
      variantCases,
      comma,
      closeBrace
    ): Item<"typeDef", VariantDef> {
      return {
        kind: "typeDef",
        boxed: {
          kind: "variant",
          name: ident.sourceString,
          cases: variantCases
            .asIteration()
            .children.map((child) => child.toModel()),
        },
      };
    },

    VariantCase(ident, lparen, type, rparen): VariantCase {
      return {
        name: ident.sourceString,
        type: type.children[0]?.toModel(),
      };
    },

    TypeItem(type_, ident, eq, typeRef, semicolon): Item<"typeDef", AliasDef> {
      return {
        kind: "typeDef",
        boxed: {
          kind: "alias",
          name: ident.sourceString,
          type: typeRef.toModel(),
        },
      };
    },

    RecordItem(
      record,
      ident,
      openBrace,
      recordFields,
      comma,
      closeBrace
    ): Item<"typeDef", RecordDef> {
      return {
        kind: "typeDef",
        boxed: {
          kind: "record",
          name: ident.sourceString,
          fields: recordFields
            .asIteration()
            .children.map((child) => child.toModel()),
        },
      };
    },

    RecordField(ident, colon, type): RecordField {
      return {
        name: ident.sourceString,
        type: type.toModel(),
      };
    },

    FlagsItems(
      flags,
      ident,
      openBrace,
      flagsFields,
      comma,
      closeBrace
    ): Item<"typeDef", FlagsDef> {
      return {
        kind: "typeDef",
        boxed: {
          kind: "flags",
          name: ident.sourceString,
          fields: flagsFields
            .asIteration()
            .children.map((child) => child.toModel()),
        },
      };
    },

    FlagsField(ident): string {
      return ident.sourceString;
    },

    NestedPackageDefinition(
      packageDecl,
      lbrace,
      packageItems,
      rbrace
    ): Item<"package", NestedPackage> {
      return {
        kind: "package",
        boxed: {
          name: packageDecl.toModel(),
          interfaces: [],
        },
      };
    },

    _iter(...children) {
      return children.map((child) => child.toModel());
    },

    ident(ident): string {
      return ident.sourceString;
    },
  });
}

export function parseWit(input: string): Wit {
  const semantics = grammar.createSemantics();
  defineModel(semantics);
  const match = grammar.match(input);
  if (match.failed()) {
    throw new Error(match.message);
  }
  return semantics(match).toModel();
}
