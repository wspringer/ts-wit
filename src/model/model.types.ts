/**
 * A utility type used for resolving extra metadata associated to WIT types.
 * Used to create data structures that are similar but extended in depth at
 * various stages of the parsing process. (This prevents use from having to
 * repeat the same structures over and over again.)
 */
type MetaOf<M, T extends string> = M extends {
  [K in T]: infer U;
}
  ? U
  : {};

export type Package<M = {}> = {
  packages: NestedPackage<M>[];
} & InterfaceContainer<M>;

export type RootPackage<M> = Package<M> & {
  name?: string;
  packages: NestedPackage<M>[];
};

export type NestedPackage<M = {}> = {
  name: string;
} & InterfaceContainer<M>;

export type Wit<M = {}> = RootPackage<M> &
  WorldContainer<M> & {
    uses: Use<M>[];
  };

export type Use<M = {}> = {
  path: string;
  alias?: string;
} & MetaOf<M, "Use">;

export type WorldContainer<M> = {
  worlds: World<M>[];
};

export type InterfaceContainer<M> = {
  interfaces: InterfaceDef<M>[];
};

export type World<M = {}> = {
  name: string;
  exports: {
    functions: Func<M>[];
    interfaces: (InterfaceDef<M> | InterfaceRef<M>)[];
  };
  imports: {
    functions: Func<M>[];
    interfaces: (InterfaceDef<M> | InterfaceRef<M>)[];
  };
  includes: Include<M>[];
  typeDefs: TypeDef<M>[];
} & Gated;

export type Include<M = {}> = {
  path: string;
  aliases?: {
    [key: string]: string;
  };
} & MetaOf<M, "Include">;

export type Func<M = {}> = {
  name: string;
  params: Param<M>[];
  result?: Ty<M>;
};

export type Param<M = {}> = {
  name: string;
  type: Ty<M>;
};

export type InterfaceDef<M = {}> = {
  name: string;
  functions: Func<M>[];
  typeDefs: TypeDef<M>[];
} & Gated &
  MetaOf<M, "InterfaceDef">;

export type InterfaceRef<M> = {
  name: string;
} & MetaOf<M, "InterfaceRef">;

export type TypeDef<M = {}> = (
  | AliasDef<M>
  | EnumDef<M>
  | VariantDef<M>
  | RecordDef<M>
  | FlagsDef<M>
) &
  MetaOf<M, "TypeDef">;

export type TupleType<M = {}> = {
  kind: "tuple";
  items: Ty<M>[];
} & MetaOf<M, "TupleType">;

export type SimpleType =
  | "s8"
  | "s16"
  | "s32"
  | "s64"
  | "u8"
  | "u16"
  | "u32"
  | "u64"
  | "f32"
  | "f64"
  | "bool"
  | "string"
  | "char";

export type RecordDef<M = {}> = {
  kind: "record";
  name: string;
  fields: RecordField<M>[];
} & MetaOf<M, "RecordDef">;

export type RecordField<M = {}> = {
  name: string;
  type: Ty<M>;
} & MetaOf<M, "RecordField">;

export type EnumDef<M = {}> = {
  kind: "enum";
  name: string;
  cases: string[];
} & MetaOf<M, "EnumDef">;

export type FlagsDef<M = {}> = {
  kind: "flags";
  name: string;
  fields: string[];
} & MetaOf<M, "FlagsDef">;

export type OptionType<M = {}> = {
  kind: "option";
  type: Ty<M>;
};

export type ListType<M = {}> = {
  kind: "list";
  type: Ty<M>;
};

export type ResultType<M = {}> = {
  kind: "result";
  ok?: Ty<M>;
  error?: Ty<M>;
};

export type VariantDef<M = {}> = {
  kind: "variant";
  name: string;
  cases: VariantCase<M>[];
} & MetaOf<M, "VariantDef">;

export type VariantCase<M = {}> = {
  name: string;
  type?: Ty<M>;
} & MetaOf<M, "VariantCase">;

export type AliasDef<M = {}> = {
  kind: "alias";
  name: string;
  type: Ty<M>;
} & MetaOf<M, "AliasDef">;

export type Ty<M = {}> =
  | TypeRef<M>
  | SimpleType
  | OptionType<M>
  | ListType<M>
  | ResultType<M>
  | TupleType<M>;

export type TypeRef<M> = {
  name: string;
} & MetaOf<M, "TypeRef">;

export type Gated = {
  since?: string;
  until?: string; // aka deprecated
  unstable?: string;
};

export type Qualified = {
  fqn: string;
};

export type Resolved<T> = {
  resolved: T | null;
};

/**
 * A qualified Wit is a Wit with all the types and interfaces qualified with
 * their FQN. (e.g. `other:foo/bar`)
 */
export type QualifiedWit = Wit<{
  TypeRef: Qualified;
  TypeDef: Qualified;
  InterfaceDef: Qualified;
  InterfaceRef: Qualified;
}>;

/**
 * A resolved Wit is a Wit with all the types and interfaces resolved to their
 * actual type definitions.
 */
export type ResolvedWit = Wit<{
  TypeRef: Resolved<TypeDef<ResolvedWit>>;
  InterfaceRef: Resolved<InterfaceDef<ResolvedWit>>;
}>;

export const test: Wit = {
  name: "foo@0.0.1",
  packages: [],
  interfaces: [
    {
      name: "foo",
      functions: [],
      typeDefs: [
        {
          kind: "record",
          name: "bar",
          fields: [
            {
              name: "message",
              type: "string",
            },
          ],
        },
      ],
    },
  ],
  uses: [
    {
      path: "other:math/operations.add",
      alias: "add",
    },
  ],
  worlds: [
    {
      name: "foo",
      since: "0.0.1",
      unstable: "foo-bar",
      exports: {
        functions: [
          {
            name: "bar",
            params: [
              {
                name: "message",
                type: {
                  kind: "option",
                  type: "string",
                },
              },
            ],
            result: {
              kind: "result",
              ok: "string",
              error: "string",
            },
          },
        ],
        interfaces: [],
      },
      imports: {
        functions: [],
        interfaces: [],
      },
      includes: [],
      typeDefs: [],
    },
  ],
};
