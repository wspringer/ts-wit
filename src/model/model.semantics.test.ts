import { parseWit } from "./model.semantics";

describe("model.semantics", () => {
  it("should parse a package declaration", () => {
    const input = "package foo:bar@0.0.1;";
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface declaration", () => {
    const input = `interface bar {
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface declaration with a parameterless function", () => {
    const input = `interface bar {
      method: func() -> string;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface declaration with a parameterized function", () => {
    const input = `interface bar {
      method: func(a: string) -> string;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with a type alias", () => {
    const input = `interface bar {
      type foo = string;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with a type alias pointing to a list", () => {
    const input = `interface bar {
      type foo = list<string>;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with a type alias pointing to an option", () => {
    const input = `interface bar {
      type foo = option<string>;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with a type alias pointing to a result", () => {
    const input = `interface bar {
      type foo = result<string, string>;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with a type alias pointing to a tuple", () => {
    const input = `interface bar {
      type foo = tuple<string, s16>;
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with an enum", () => {
    const input = `interface bar {
      enum foo {
        A,
        B,
        C,
      }
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse an interface with a variant", () => {
    const input = `interface bar {
      variant foo {
        A(string),
        B,
        C,
      }
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse a record", () => {
    const input = `interface bar {
      record foo {
        a: string,
        b: int,
      }
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse flags", () => {
    const input = `interface bar {
      flags foo {
        A, B, C
      }
    }`;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse nested packages", () => {
    const input = `
    package foo:bar@0.0.1 {
      interface baz {
        method: func() -> string;
      }
    }
    `;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse a world", () => {
    const input = `
    world foo {
    }
    `;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse a world with an exported function", () => {
    const input = `
    world foo {
      export bar: func() -> string;
    }
    `;
    expect(() => parseWit(input)).not.toThrow();
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse a world with an exported interface", () => {
    const input = `
    world foo {
      export bar: interface {}
    }
    `;
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });

  it("should parse a world with an interface with members ", () => {
    const input = `
    world foo {
      export bar: interface {
        method: func() -> string;
      }
    }
    `;
    const parsed = parseWit(input);
    expect(parsed).toMatchSnapshot();
  });
});
