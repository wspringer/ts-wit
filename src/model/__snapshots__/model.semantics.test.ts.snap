// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model.semantics should parse a package declaration 1`] = `
{
  "interfaces": [],
  "name": "foo:bar@0.0.1",
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse a record 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "fields": [
            {
              "name": "a",
              "type": "string",
            },
            {
              "name": "b",
              "type": "int",
            },
          ],
          "kind": "record",
          "name": "foo",
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse a world 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse a world with an exported function 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [
          {
            "name": "bar",
            "params": [],
            "result": "string",
          },
        ],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse a world with an exported interface 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [
          {
            "functions": [],
            "name": "bar",
            "typeDefs": [],
          },
        ],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse a world with an imported function 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [
          {
            "name": "bar",
            "params": [],
            "result": "string",
          },
        ],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse a world with an imported interface 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [
          {
            "functions": [],
            "name": "bar",
            "typeDefs": [],
          },
        ],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse a world with an interface with members  1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [
          {
            "functions": [
              {
                "name": "method",
                "params": [],
                "result": "string",
              },
            ],
            "name": "bar",
            "typeDefs": [],
          },
        ],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse an interface declaration 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface declaration with a parameterized function 1`] = `
{
  "interfaces": [
    {
      "functions": [
        {
          "name": "method",
          "params": [
            {
              "name": "a",
              "type": "string",
            },
            [],
            [],
          ],
          "result": "string",
        },
      ],
      "name": "bar",
      "typeDefs": [],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface declaration with a parameterless function 1`] = `
{
  "interfaces": [
    {
      "functions": [
        {
          "name": "method",
          "params": [],
          "result": "string",
        },
      ],
      "name": "bar",
      "typeDefs": [],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with a type alias 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "kind": "alias",
          "name": "foo",
          "type": "string",
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with a type alias pointing to a list 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "kind": "alias",
          "name": "foo",
          "type": {
            "kind": "list",
            "type": "string",
          },
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with a type alias pointing to a result 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "kind": "alias",
          "name": "foo",
          "type": {
            "error": undefined,
            "kind": "result",
            "ok": "string",
          },
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with a type alias pointing to a tuple 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "kind": "alias",
          "name": "foo",
          "type": {
            "items": [
              "string",
              "s16",
            ],
            "kind": "tuple",
          },
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with a type alias pointing to an option 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "kind": "alias",
          "name": "foo",
          "type": {
            "kind": "option",
            "type": "string",
          },
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with a variant 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "cases": [
            {
              "name": "A",
              "type": "string",
            },
            {
              "name": "B",
              "type": undefined,
            },
            {
              "name": "C",
              "type": undefined,
            },
          ],
          "kind": "variant",
          "name": "foo",
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with an enum 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "cases": [
            "A",
            "B",
            "C",
          ],
          "kind": "enum",
          "name": "foo",
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse an interface with an enum 2`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "cases": [
            "A",
            "B",
            "C",
          ],
          "kind": "enum",
          "name": "foo",
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse deprecated gates in a world 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
      "until": "1.0.0",
    },
  ],
}
`;

exports[`model.semantics should parse flags 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [
        {
          "fields": [
            "A",
            "B",
            "C",
          ],
          "kind": "flags",
          "name": "foo",
        },
      ],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse gates in a world 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [],
      "unstable": "jazz",
    },
  ],
}
`;

exports[`model.semantics should parse gates on interfaces 1`] = `
{
  "interfaces": [
    {
      "functions": [],
      "name": "bar",
      "typeDefs": [],
      "unstable": "jazz",
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse gates on interfaces with members 1`] = `
{
  "interfaces": [
    {
      "functions": [
        {
          "name": "method",
          "params": [],
          "result": "string",
          "unstable": "jazz",
        },
      ],
      "name": "bar",
      "typeDefs": [],
    },
  ],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse include statements 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [
        {
          "path": "foo:bar@0.0.1",
        },
      ],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse include statements with aliases 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [
        {
          "aliases": {
            "baz": "qux",
          },
          "path": "foo:bar@0.0.1",
        },
      ],
      "name": "foo",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse nested packages 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [
    {
      "interfaces": [],
      "name": "foo:bar@0.0.1",
    },
  ],
  "uses": [],
  "worlds": [],
}
`;

exports[`model.semantics should parse since gates in a world 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "since": "1.0.0",
      "typeDefs": [],
    },
  ],
}
`;

exports[`model.semantics should parse top level use statements 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [
    {
      "alias": undefined,
      "path": "foo:bar@0.0.1",
    },
  ],
  "worlds": [],
}
`;

exports[`model.semantics should parse top level use statements with an alias 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [
    {
      "alias": "baz",
      "path": "foo:bar@0.0.1",
    },
  ],
  "worlds": [],
}
`;

exports[`model.semantics should parse type defs in a world 1`] = `
{
  "interfaces": [],
  "name": undefined,
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [],
      },
      "imports": {
        "functions": [],
        "interfaces": [],
      },
      "includes": [],
      "name": "foo",
      "typeDefs": [
        {
          "kind": "alias",
          "name": "bar",
          "type": "string",
        },
      ],
    },
  ],
}
`;
